states/module.go

```diff
@@ -88,23 +88,56 @@ func (ms *Module) RemoveResource(addr addrs.Resource) {
 // are updated for all other instances of the same resource as a side-effect of
 // this call.
 func (ms *Module) SetResourceInstanceCurrent(addr addrs.ResourceInstance, obj *ResourceInstanceObjectSrc, provider addrs.AbsProviderConfig) {
-	ms.SetResourceMeta(addr.Resource, eachModeForInstanceKey(addr.Key), provider)
-
 	rs := ms.Resource(addr.Resource)
-	is := rs.EnsureInstance(addr.Key)
-
-	is.Current = obj
-
-	if !is.HasObjects() {
-		// If we have no objects at all then we'll clean up.
-		delete(rs.Instances, addr.Key)
+	// if the resource is nil and the object is nil, don't do anything!
+	// you'll probably just cause issues
+	if obj == nil && rs == nil {
+		return
 	}
-	if rs.EachMode == NoEach && len(rs.Instances) == 0 {
-		// Also clean up if we only expect to have one instance anyway
-		// and there are none. We leave the resource behind if an each mode
-		// is active because an empty list or map of instances is a valid state.
-		delete(ms.Resources, addr.Resource.String())
+	if obj == nil && rs != nil {
```

Comment:
On the "old" side off this diff I see it has an extra guard `rs.EachMode == NoEach`, but I don't see a corresponding guard in the new codepath. Did something else change in the meantime to make that unnecessary?

My memory of this (which, sadly, is rusty) is that the idea here was to preserve the resource object for a little while longer when `count` becomes zero (after formerly being >0) or `for_each` becomes `{}` (after formerly being non-empty), so that later references to the resource can still return an empty list or empty map correctly. We then have a separate graph node in the apply graph which aims to finally clean this up by calling `ms.RemoveResource`:

https://github.com/hashicorp/terraform/blob/6e909e3e8468599fb6eb32ff81fa8df750331e2a/terraform/graph_builder_apply.go#L108-L119

If I'm remembering the situation correctly, I think we could test it with a configuration something like this:

```hcl
variable "a_count" {
  type = number
}

resource "test" "a" {
  count = var.a_count
}

resource "test" "b" {
  foo = test.a.*.id
}
```

If we were to plan/apply the above initially with `a_count` set to `2` and then again with `a_count` set to `0` then hopefully we'd see `foo` inside `test.b` correctly set to an empty list on the second apply. If it were not working, I think it might turn up as an unknown value or some other error, depending on how the expression evaluation code handles not finding `test.a` in the state at all.

In particular, I'm worried about this part of the evaluator logic where it handles the resource state coming back as `nil`, which I think is what would happen if we remove it from the state too early:

https://github.com/hashicorp/terraform/blob/e32641c9ce8b7c1b766f7d0368a7aec93d3438ce/terraform/evaluate.go#L542-L552

Reply:
A further thought here, after some additional pondering:

I wonder if perhaps the situation we were trying to allow for here was the evaluation of the dependent expression (`test.a.*.id` in the above example) during the _plan_ phase. Testing for that in _automated_ testing is likely to be fiddly (though possible, I think) but perhaps a quick _manual_ test would be to work with the configuration I described above but look closely at the `terraform plan` output on the second plan (when `a_count` is set to zero for the first time).

Ideally we'd want to see something like this:

```hcl
    ~ foo: [
        - "some-id-1",
        - "some-id-2",
      ]
```

...but I worry we might see something like this instead:

```hcl
    ~ foo: [
          "some-id-1",
          "some-id-2",
      ] -> (known after apply)
```

It being briefly unknown isn't a big deal in this example, but the difference between that being known-empty or unknown can be important if there are any expressions deriving in turn from `test.a.foo`, which could make this a breaking change.

---

I wrote the above example originally thinking about it going in a context test with a mocked provider that could provide that `foo` argument, but to test it in a manual testing situation would require using a "real" provider. Hopefully that example translates okay to using `null_resource`, though since null resource `triggers` only accepts strings it'll require some additional complexity I suppose:

```hcl
  triggers = {
    foo = join(",", null_resource.a.*.id)
  }
```

This'll mean that the "good" diff would be from a non-empty string to an empty string, rather than a non-empty list to an empty list, but a similar principle should apply.

Reply:
Thanks for these! The manual testing did indeed confirm this is an issue ... `master` puts a reference error (as it should) but this branch does not. Will keep re-reading these comments and try and develop some understanding.

One thing I'm going to need to read (I did a bunch of poking at the manual test before reading rest of the comment and the second comment) is understanding why evaluate is thinking differently about `nil` resources.

Presently, what happens on the branch is that it will go through `plan` but then there are unknown values during apply. The reference issue should be caught during plan, so that's what I'm trying to run down.

Reply:
Got it, I think. That the NoEach check can happen in this function, so that we don't delete the resource early (I added a similar comment to what you wrote, that we allow resources with 0 instances to continue to exist if they're a Map or List). Now to wait for the tests to come back 🙃

Reply:
It was the tests that I'd updated! 😬 both a good and bad feeling.